File: BibliotecaUFABC/app/src/main/java/com/nintersoft/bibliotecaufabc/webviewclients/LoginWebClient.java
package com.nintersoft.bibliotecaufabc.webviewclients;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Build;
import android.webkit.ValueCallback;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.nintersoft.bibliotecaufabc.activities.LoginActivity;
import com.nintersoft.bibliotecaufabc.utilities.GlobalConstants;
import com.nintersoft.bibliotecaufabc.utilities.GlobalFunctions;
import com.nintersoft.bibliotecaufabc.utilities.GlobalVariables;

import androidx.annotation.RequiresApi;

public class LoginWebClient extends WebViewClient {
    private int login_page_finished;
    private int login_services_finished;

    private Context mContext;

    public LoginWebClient(Context context){
        super();
        mContext = context;
        login_page_finished = 0;
        login_services_finished = 0;
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
        return false;
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        return false;
    }

    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
        super.onPageStarted(view, url, favicon);
    }

    @Override
    public void onPageFinished(WebView view, String url) {
        super.onPageFinished(view, url);

        if (url.contains(GlobalConstants.URL_LIBRARY_LOGIN_P)){
            if (login_page_finished == 0){
                if (GlobalVariables.storeUserFormData)
                    ((LoginActivity)mContext).setSavedUserLogin();
                else
                    ((LoginActivity)mContext).setupInterface(true);
                login_page_finished++;
            }
        }
        else if (url.contains(GlobalConstants.URL_LIBRARY_SERVICES)){
            if (login_services_finished < 1){
                login_services_finished++;
                return;
            }

            String script = String.format("%1$s \ngetUsername();",
                    GlobalFunctions.getScriptFromAssets(mContext, "javascript/login_scraper.js"));
            view.evaluateJavascript(script, new ValueCallback<String>() {
                @Override
                public void onReceiveValue(final String value) {
                    ((LoginActivity)mContext).runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            if (!value.isEmpty()) ((LoginActivity)mContext).hasLoggedIn(value);
                        }
                    });
                }
            });
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    public void onReceivedError(WebView view, WebResourceRequest request, final WebResourceError error) {
        super.onReceivedError(view, request, error);

        ((LoginActivity)mContext).runOnUiThread(new Runnable() {
            @Override
            public void run() {
                ((LoginActivity)mContext).setErrorForm(error.getDescription().toString());
            }
        });
    }

    @Override
    public void onReceivedError(WebView view, int errorCode, final String description, String failingUrl) {
        super.onReceivedError(view, errorCode, description, failingUrl);

        ((LoginActivity)mContext).runOnUiThread(new Runnable() {
            @Override
            public void run() {
                ((LoginActivity)mContext).setErrorForm(description);
            }
        });
    }
}

File: BibliotecaUFABC/app/src/main/java/com/nintersoft/bibliotecaufabc/webclient/LoginWebClient.kt
package com.nintersoft.bibliotecaufabc.webclient

import android.os.Build
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.annotation.RequiresApi
import androidx.lifecycle.LiveData
import androidx.preference.PreferenceManager
import com.nintersoft.bibliotecaufabc.R
import com.nintersoft.bibliotecaufabc.global.Constants
import com.nintersoft.bibliotecaufabc.global.Functions
import com.nintersoft.bibliotecaufabc.model.AppContext
import com.nintersoft.bibliotecaufabc.ui.search.SingleLiveEvent


class LoginWebClient : WebViewClient() {

    private val mContext = AppContext.context!!
    private var loginPageFinished = 0
    private var loginServicesFinished = 0
    private val _error = SingleLiveEvent<Boolean?>()
    private val _hasLoggedIn = SingleLiveEvent<String?>()
    private val _loginRequest = SingleLiveEvent<Boolean?>()

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
        return false
    }

    override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean { return false }

    override fun onPageFinished(view: WebView?, url: String?) {
        super.onPageFinished(view, url)
        if (url.isNullOrEmpty()) return

        if (url.contains(Constants.URL_LIBRARY_LOGIN_P) && loginPageFinished == 0){
            _loginRequest.value = PreferenceManager.getDefaultSharedPreferences(mContext).
                getBoolean(mContext.getString(R.string.key_privacy_store_password), true)
            loginPageFinished++
        }
        else if (url.contains(Constants.URL_LIBRARY_SERVICES)){
            if (loginServicesFinished < 1){
                loginServicesFinished++
                return
            }

            val script = "${Functions.scriptFromAssets("js/login_scp.js")} \ngetUsername();"
            view?.evaluateJavascript(script, ({ _hasLoggedIn.value = it }))
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onReceivedError(view: WebView?, request: WebResourceRequest?, error: WebResourceError?) {
        super.onReceivedError(view, request, error)
        loginPageFinished = 0
        loginServicesFinished = 0
        _error.value = true
    }

    @Suppress("DEPRECATION")
    override fun onReceivedError(view: WebView?, errorCode: Int, description: String?, failingUrl: String?) {
        super.onReceivedError(view, errorCode, description, failingUrl)
        loginPageFinished = 0
        loginServicesFinished = 0
        _error.value = true
    }

    fun errorEvent() : LiveData<Boolean?>{ return _error }
    fun loginStatus() : LiveData<String?> { return _hasLoggedIn }
    fun loginRequested() : LiveData<Boolean?> { return _loginRequest }
}